### PR TITLE
[RunAllTests] Fix #2900: Add network interceptor to request headers

### LINKED ISSUE #2900: Implement network interceptor to add to outgoing request header
Fix part of #53
Implement a network interceptor to catch outgoing requests and add a header with the API key, package version, and package name, and version code for receiving client to authenticate the request 

[Design doc](https://docs.google.com/document/d/1_RTzad6u4m-ZFn0VTWNddp4bc0fCYrMk-ABIVYVJ7gU/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by jcqli**
## Explanation
Fix part of #53
Fix #2900
Implements a new network interceptor that adds request headers to outgoing requests as part of feedback reporting project, although will be applied to all outgoing messages. Renames original `NetworkInterceptor` => `JsonPrefixNetworkInterceptor` to create separate interceptors with independent roles.

[Design doc](https://docs.google.com/document/d/1_RTzad6u4m-ZFn0VTWNddp4bc0fCYrMk-ABIVYVJ7gU/edit?usp=sharing)

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by anandwana001** [comment]
@jcqli Setting the script will help you prevent kotlin lint failures. - https://github.com/oppia/oppia-android/wiki#installation

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: This doesn't seem to set the versionName or versionCode -- it throws a NullPointerException later on in the test for the versionName when the RemoteAuthNetworkInterceptor tries to set the headers ([line 44)](https://github.com/oppia/oppia-android/pull/2907/files#diff-c443970e72f2a145c0321816f33ec349b2cdf2054e8bc3b2f0338cb742ef2cf2R44). Any tips?

Comment (BenHenning) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: I think NPE can happen in Context.getVersionName() if there's no PackageInfo for the current package name. You should be able to debug in this test what the package name is for Context and whether it's matching the PackageInfo you're creating (which is using package name 'test_package_name').

File: utility/src/main/java/org/oppia/android/util/extensions/BUILD.bazel
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/extensions/BUILD.bazel:19]: Nice catch.

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Should be on the same line.

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: These should be parameter injected, not constructed, otherwise their injections won't be ready.

You can do that by adding these as parameters to the provider method.

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: You should remove this--immutable strings don't need to be scoped since they can be trivially re-retrieved.

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]:
```suggestion
  // replaced with a secret key in production.
```

File: data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt]: This class is stateless so it shouldn't be scoped.

File: data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt]: Can this be private? It seems to only be needed for the class implementation.

File: data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt:30]: Please also inject this everywhere it's used in tests, too.

Though, it might be helpful to wait for #2642 if we can since I think that's largely simplifying data test setup.

File: utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt]:
```suggestion
  return this.packageManager.getPackageInfo(this.packageName, /* flags= */ 0).versionCode
```

Maybe fix the one above, too, for an opportunistic fix.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jcqli! Left some thoughts. Not sure if you actually want to block on #2916 (might be better not to give the time sensitivity here), but wanted to mention it as an option.

File: data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt]: Is there a way to declare this as test-accessible? The network interceptor tests use this to verify the correct headers are added.

Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt]: I made this an internal fun so that the associated test could call on it and verify that headers are added correctly. I think another way to test this and keep the function private would be calling `intercept()` on a chain and verifying the response's associated request matched the interceptor's changes, but this is a bit more indirect (and probably more difficult) and I'm not sure if it's an accurate representation of the request anymore after it goes through the request. Using an internal function seems to be the most beneficial, WDYT?

File: data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt]: Done

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Done

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Done

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Done

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Done, I saw this as well but it didn't show up on my AS until now. How odd

File: utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt
Comment (jcqli) [reply utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt]: Done

File: data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt:30]: Is this interceptor needed for the mock retrofit? All the service tests pass when removing it, maybe that's an indicator that our mock retrofit is not set up properly.

I ask since injecting the interceptors means we then need to declare TestApplications for all the tests that use the interceptors, is that right?

**Comment by jcqli** [comment]
> Thanks @jcqli! Left some thoughts. Not sure if you actually want to block on #2916 (might be better not to give the time sensitivity here), but wanted to mention it as an option.

I think I'd rather have this merged and not blocked on #2916 but would need to make some changes so that its' not blocked by #2878, let me know if that would be better (I think it would make more sense, but requires some more work on my end).

Unrelated, I'm not sure why the linter is giving errors on the third_party/versions.bzl file (from 2878 changes). I think it's about how I added the MockWebServer to dependency but this can be resolved in the other PR, and also depends on if we block on 2878. Let me know your thoughts.

File: data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt:30]: Based on our discussion offline I'm going to leave the interceptor here so that it doesn't interfere with #2642 since it seems that the interceptors will be added there.

File: third_party/versions.bzl
Comment (prayutsu) [inline third_party/versions.bzl:89]: I think lint is failing because there's a space missing before the version (after the colon).

**Comment by jcqli** [comment]
To properly test the network interceptor without exposing private functions (e.g. actually testing the full network request was successful), we used MockWebServer to verify the request call. Turns out, MockWebServer is incompatible with MockRetrofit so the test here does not use the mock services but rather the full service on the mocked server.

/cc @prayutsu Not sure if this will affect your data module PR but might be relevant

File: data/src/test/java/org/oppia/android/data/backends/gae/JsonPrefixNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/JsonPrefixNetworkInterceptorTest.kt]: This should be removed--only the test needs an injector generated.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:73]: I suggest moving your test helpers to also be defined in the same order they're called here (& after the tests since we generally follow an "appendix" style organization for helpers rather than "footnote" style).

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Suggest going with 10 seconds for parity with TestCoroutineDispatcher (in fact you could use the fields from there for the time & unit if you'd prefer).

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]:
```suggestion
    verifyRequestHeaders(interceptedRequest?.headers)
```

Maybe use nullable calls, instead. The assertion should still fail if the individual retrievals are null, though also recommending a non-null assertion for the request headers in verifyRequestHeaders to make failures a bit more explicit.

'!!' isn't quite as ideal since it will throw a NPE rather than AssertionError. It'll still be clear once someone digs into the failure, but test frameworks generally expect AssertionErrors and may provide slightly better error messaging in those cases. Plus, it's just nice to never use '!!' :)

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: This isn't doing anything--I suggest removing it.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Should this be initialized in setUpRetrofit, instead?

Also, does it need to be torn down in an ``@After`` function?

Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Should this be initialized in setUpRetrofit, instead?

Also, does it need to be torn down in an ``@After`` function?

File: data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt
Comment (BenHenning) [reply data/src/main/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptor.kt]: It looks like this is resolved now.

File: data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt
Comment (BenHenning) [reply data/src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest.kt:30]: Moved this comment to #2916 via https://github.com/oppia/oppia-android/pull/2916/files#r610890818.

File: third_party/versions.bzl
Comment (BenHenning) [reply third_party/versions.bzl:89]: Nice catch.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jcqli. Not sure why my previous review was submitted automatically, but either way--just have a few comments to address but otherwise I think the PR looks good.

**Comment by BenHenning** [comment]
@jcqli you may also want to assign the other code owners so that they can take a look.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: @BenHenning I realized that this test doesn't actually use the request with the incorrectly set headers. I don't believe MockWebServer enables us to enqueue a specific request, nor do I know how to pass a request directly into the service (unless we want to mock the service differently). Do you have any suggestions on how we could test this case?

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Done

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:73]: Ack, done

File: data/src/test/java/org/oppia/android/data/backends/gae/JsonPrefixNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/JsonPrefixNetworkInterceptorTest.kt]: Done

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Ack, done

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Ah oops thanks! Done

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Yup, done

Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Done

**Comment by jcqli** [comment]
> @jcqli you may also want to assign the other code owners so that they can take a look.

Ack, done!

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkApiKey.kt
Comment (rt4914) [inline data/src/main/java/org/oppia/android/data/backends/gae/NetworkApiKey.kt]: Nit: `.` at the end.

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (rt4914) [inline data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Suggest changing this code to either one of these
```
    val clientBuilder = OkHttpClient.Builder()
      .addInterceptor(jsonPrefixNetworkInterceptor)
      .addInterceptor(remoteAuthNetworkInterceptor)
```

or 

```
    val client = OkHttpClient.Builder()
      .addInterceptor(jsonPrefixNetworkInterceptor)
      .addInterceptor(remoteAuthNetworkInterceptor)
      .build()
```

Because currently your object `client` is actually a `clientBuilder`

File: data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt
Comment (rt4914) [inline data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt:32]: Does `FeedbackReportingService` do not need `JsonPrefixNetworkInterceptor ` ?

**Review by rt4914 — APPROVED** [review]
LGTM, there are three comments. Two you can resolve and other one you can answer and I will check it. Thanks.

**Review by anandwana001 — APPROVED** [review]
test files LGTM

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: I was thinking more like: verifyRequestedHeaders(it?.headers). The null check should work automatically inside verify since you'll also need to do the safe-call operator & Truth will fail if the resulting expression evaluates to null.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:103]: Maybe explain briefly via comment why we can't use Retrofit in this test case.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (BenHenning) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Per offline discussion I think we landed on using Okhttp directly (& per the latest code)--can this be resolved now?

**Review by BenHenning — COMMENTED** [review]
Thanks @jcqli! Generally LGTM, but can you re-run the Bazel tests? I think you're running into one of the known CI issues where compute_affected_tests hangs. You'll need to add '[RunAllTests]' to the PR title & then re-run the Bazel tests workflow.

File: data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt:32]: The service currently doesn't receive anything in the response, so I don't believe so since the interceptor only handles responses.

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/NetworkModule.kt]: Done

File: data/src/main/java/org/oppia/android/data/backends/gae/NetworkApiKey.kt
Comment (jcqli) [reply data/src/main/java/org/oppia/android/data/backends/gae/NetworkApiKey.kt]: Done

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Ah yup, thanks for following-up on this

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:103]: Done

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt
Comment (jcqli) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt]: Done

**Review by BenHenning — COMMENTED** [review]
Thanks @jcqli! Looks like you need to update TestDispatcherModule & RobolectricModule imports in your new tests--the builds for those tests are failing on both Bazel & Gradle. In general, it's good practice to re-run key tests locally (such as new/changed suites) after a merge just to double-check integrations post-sync.

PR LGTM otherwise.

File: data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt
Comment (rt4914) [reply data/src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest.kt:32]: Ack. Thanks.

**Comment by jcqli** [comment]
> Thanks @jcqli! Looks like you need to update TestDispatcherModule & RobolectricModule imports in your new tests--the builds for those tests are failing on both Bazel & Gradle. In general, it's good practice to re-run key tests locally (such as new/changed suites) after a merge just to double-check integrations post-sync.

@BenHenning Thanks for taking a look!

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jcqli. ``RemoteAuthNetworkInterceptorTest`` appears broken on Bazel. Please fix & send back. (I highly encourage you verify that test builds & passes locally before sending back since the CI run will take ~2 hours to run).

**Comment by jcqli** [comment]
> Thanks @jcqli. `RemoteAuthNetworkInterceptorTest` appears broken on Bazel. Please fix & send back. (I highly encourage you verify that test builds & passes locally before sending back since the CI run will take ~2 hours to run).

Aw man, I missed the data module -- thanks for the pointer, it should be good now!

**Review by BenHenning — APPROVED** [review]
Looks like everything is passing now--LGTM. Thanks @jcqli!

**Pull request merged**
